### PR TITLE
convert sicmutils.rational-function to cljs

### PR DESCRIPTION
**Pull request by sritchie**
All set! I'll leave a review pointing out some subtleties we want to consider.

### CODE REVIEW

File: src/sicmutils/rational_function.cljc [hunk 1]
```diff
@@ -39,23 +40,46 @@
...
+       (toString [p] (str u " : " v))
+       (equals [_ b]
+               (and (instance? RationalFunction b)
+                    (and (= arity (.-arity b))
+                         (= u (.-u b))
+                         (= v (.-v b)))))]
+
+      :cljs
+      [Object
+       (toString [p] (str u " : " v))
```
Comment (sritchie) [inline src/sicmutils/rational_function.cljc:54]: @littleredcomputer , what do you think of this `toString` implementation?

File: src/sicmutils/rational_function.cljc [hunk 2]
```diff
@@ -39,23 +40,46 @@
...
+                   (.toString x)
+                   "\"]"))]))
+
+(defn rational-function? [r]
+  (instance? RationalFunction r))
 
 (defn make
   "Make the fraction of the two polynomials p and q, after dividing
   out their greatest common divisor."
-  [^Polynomial u ^Polynomial v]
```
Comment (sritchie) [inline src/sicmutils/rational_function.cljc:53]: these type hints weren't buying us anything, so I removed them for now. I DO miss all of this as documentation, btw. In a past life I used https://github.com/plumatic/schema to document everything, like this:

```clojure
(s/defn exception-string :- String
  "Returns a nice, pretty string (with good newlines) of the supplied
  throwable."
  [e :- Throwable]
  (binding [ex/*fonts* nil]
    (ex/format-exception e)))
```

that would be a nice thing to add here.

File: src/sicmutils/rational_function.cljc [hunk 3]
```diff
@@ -67,7 +91,7 @@
         cs (into (into #{} cv) (p/coefficients u))
         integerizing-factor (*
                              (if (< lcv 0) -1 1)
-                             (reduce euclid/lcm 1 (map denominator (filter ratio? cs))))
+                             (reduce euclid/lcm 1 (map u/denominator (filter u/ratio? cs))))
```
Comment (sritchie) [inline src/sicmutils/rational_function.cljc:94]: this sets us up for this all to work once I get fractions implemented.

File: src/sicmutils/rational_function.cljc [hunk 4]
```diff
@@ -243,63 +265,33 @@
...
-  [c ^RationalFunction r]
-  (make (g/mul c (.-u r)) (.-v r)))
+(defmethod g/sub [::rational-function ::rational-function] [a b] (sub a b))
+(defmethod g/sub [::rational-function ::p/polynomial] [r p] (subp r p))
 
-(defmethod g/mul
-  [::rational-function Double]
-  [^RationalFunction r c]
-  (make (g/mul (.-u r) c) (.-v r)))
+(defmethod g/sub [::rational-function ::v/integral] [r c]
```
Comment (sritchie) [inline src/sicmutils/rational_function.cljc:282]: this is a harder diff to read, but I basically backed us out to `::v/integral` in many spots, and `::v/number` where you wanted support for doubles. I think some of the implementations are missing. We can fill those in with better tests, I think.

Would you mind viewing the file (outside of the diff) and letting me know if you think we've laid this out like we want?

File: src/sicmutils/rational_function.cljc (hunk 1 above)
Comment (littleredcomputer) [reply src/sicmutils/rational_function.cljc:54]: SG, ":" for ratios is fine

File: src/sicmutils/rational_function.cljc (hunk 4 above)
Comment (littleredcomputer) [reply src/sicmutils/rational_function.cljc:282]: This I have done, and I liked it fine

**Pull request merged**